va-memorable-date, va-select - Accessibility updates from staging review

powellkerry opened the pull request:
## Chromatic
<!-- This `{{head.ref}}` is a placeholder for a CI job - it will be updated automatically -->
https://{{head.ref}}--60f9b557105290003b387cd5.chromatic.com

## Description
- va-memorable-date: move error text inside of legend to ensure that it is read by screen readers
- va-select: Fix bug with aria-describedby where the element id was incorrect for the USWDS version, added a fallback to the help text if it is defined

Closes https://github.com/department-of-veterans-affairs/va.gov-team/issues/64284

## Testing done
- Local testing via chrome and Mac VoiceOver

## Screenshots
- No visual changes

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

briandeconinck commented:
@powellkerry Testing with NVDA, there's something that's preventing the error message from being read when tabbing into the memorable date component. Looking at it with @jeana-adhoc it looks like it might be something connected to the `role="alert"` on the error message, which is a little concerning. I'm going to try to dig into this a bit next week to see if we can figure out the right way forward here.

cc @rsmithadhoc

rsmithadhoc commented:
I looked into getting the error to read when re-tabbing to the field. I wasn't able to get it to work with the current markup.

One possible idea is utilizing the input field's error message, possibly making it sr-only so it doesn't misalign the fields. I don't know if that is a great solution.
<img width="1792" alt="image" src="https://github.com/department-of-veterans-affairs/component-library/assets/100248909/8ebdc192-940a-4c0a-a049-a48f9804c9fd">

I'll try some more things today and let you know if I come up with anything better.

rsmithadhoc commented:
@powellkerry The other way I was able to get this to read out was by setting `shadow: false` on `va-text-input` and putting `aria-describedby="error-message"` on the inner `<input>` tag itself. It seems like other methods are not able to see outside of the `va-text-input` shadow DOM and read out that error message. Using `aria-describedby` with an ID of the error message is the way I've seen it done before, and that works perfectly in plain HTML, but our setup complicates this.

<img width="1792" alt="image" src="https://github.com/department-of-veterans-affairs/component-library/assets/100248909/30f32257-facd-4866-b185-83c193db5de9">
<img width="1792" alt="image" src="https://github.com/department-of-veterans-affairs/component-library/assets/100248909/a8b43026-b8d6-43f3-90cc-5b8eebaca9f1">

However, I'm not sure if it is feasible to disable the shadow DOM for the input (or if it is even possible to do it as a one-off for memorable date). It definitely had some styling issues. Because of that, I think my first comment still may be the better choice, unless you know of any workaround.

powellkerry commented:
@rsmithadhoc @jamigibbs I have updated memorable date to pass validation errors to the individual va-text-input error props. It seems like a bit of hack, but I also added the `sr-only` class on each of the va-text-input fields' errors when the memorable date component loads. In my testing the error is now read on each invalid field when focussed.

jamigibbs commented on packages/web-components/src/components/va-memorable-date/va-memorable-date.tsx:
```diff
@@ -192,6 +192,14 @@ export class VaMemorableDate {
+    const textInputs = this.el.shadowRoot.querySelectorAll('va-text-input');
+    textInputs.forEach((input) => {
+      input.shadowRoot.querySelector('#input-error-message').classList.add('sr-only');
+    });
```
I'm ok with this approach. I think trying to add a conditional `sr-only` directly in the child components would make things more complicated than they need to be.

I just want to double check that we don't need to do this also with va-select when the month dropdown is used?

powellkerry replied:
I have added this solution to the month drop down, it also seems that validation was failing on the month drop down. I have added a quick fix for that so that the value of "0" will not pass validation.

rsmithadhoc commented:
@powellkerry Definitely seems hacky, but might be the only way due to the shadow DOM. 

I tested with NVDA, it worked perfectly. But with VoiceOver, it was also reading the error on the month field (for an error with day). It is picking that up from inside the `<legend>`. So I think the error display under the hint text may need to be outside of the `<legend>`.
<img width="1134" alt="image" src="https://github.com/department-of-veterans-affairs/component-library/assets/100248909/72f6368b-0769-4faa-a26a-0c5b02d23637">

<img width="1792" alt="image" src="https://github.com/department-of-veterans-affairs/component-library/assets/100248909/72cee8e9-7eb6-4ec3-80fc-c6d1e6e7fd4f">

Other than that, I think the solution looks good.

powellkerry commented:
@rsmithadhoc I moved the error span back outside of the legend element. Let me know if that is better or worse.

rsmithadhoc reviewed (approved):
Looks good, thanks!
